### PR TITLE
[SPARK-11432][GraphX] Personalized PageRank shouldn't use uniform initialization

### PR DESCRIPTION
**Pull request by moustaki**
Changes the personalized pagerank initialization to be non-uniform.

### CODE REVIEW

**Comment by dbtsai** [comment]
Jenkins, add to whitelist.

**Comment by dbtsai** [comment]
Jenkins, ok to test.

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 1]
```diff
@@ -104,17 +104,25 @@ object PageRank extends Logging {
       graph: Graph[VD, ED], numIter: Int, resetProb: Double = 0.15,
       srcId: Option[VertexId] = None): Graph[Double, Double] =
   {
+    val personalized = srcId isDefined
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: Add extra empty line after this.

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 2]
```diff
@@ -104,17 +104,25 @@ object PageRank extends Logging {
...
       .mapTriplets( e => 1.0 / e.srcAttr, TripletFields.Src )
       // Set the vertex attributes to the initial pagerank values
-      .mapVertices( (id, attr) => resetProb )
+      .mapVertices( (id, attr) => {
+        if (personalized) {
+          if (id == srcId.get) resetProb else 0.0
+        } else {
+          resetProb
+        }
+      })
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]:
```scala
      .mapVertices { (id, attr) => 
        if (personalized) {
          if (id == srcId.get) resetProb else 0.0
        } else {
          resetProb
        }
      }
```

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 3]
```diff
@@ -192,6 +200,7 @@ object PageRank extends Logging {
       graph: Graph[VD, ED], tol: Double, resetProb: Double = 0.15,
       srcId: Option[VertexId] = None): Graph[Double, Double] =
   {
+    val personalized = srcId.isDefined
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: ditto

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 4]
```diff
@@ -202,10 +211,15 @@ object PageRank extends Logging {
       // Set the weight on the edges based on the degree
       .mapTriplets( e => 1.0 / e.srcAttr )
       // Set the vertex attributes to (initalPR, delta = 0)
-      .mapVertices( (id, attr) => (0.0, 0.0) )
+      .mapVertices( (id, attr) => {
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: ditto

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 5]
```diff
@@ -104,17 +104,25 @@ object PageRank extends Logging {
...
+        if (personalized) {
+          if (id == srcId.get) resetProb else 0.0
+        } else {
+          resetProb
+        }
+      })
 
-    val personalized = srcId isDefined
     val src: VertexId = srcId.getOrElse(-1L)
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: Move up `val src: VertexId = srcId.getOrElse(-1L) ` to `val personalized = srcId isDefined`

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala (hunk 2 above)
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: With src, let's do

```scala
      .mapVertices { (id, attr) => 
          if (id == src) resetProb else 0.0
      }
```

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 6]
```diff
@@ -202,10 +211,15 @@ object PageRank extends Logging {
...
       .mapTriplets( e => 1.0 / e.srcAttr )
       // Set the vertex attributes to (initalPR, delta = 0)
-      .mapVertices( (id, attr) => (0.0, 0.0) )
+      .mapVertices( (id, attr) => {
+        if (personalized && id == srcId.get) {
+          (resetProb, Double.NegativeInfinity)
+        } else {
+          (0.0, 0.0)
+        }
+      } )
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: ditto~

```scala
if (id == src) (resetProb, Double.NegativeInfinity) else (0.0, 0.0)
```

```

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 7]
```diff
@@ -202,10 +211,15 @@ object PageRank extends Logging {
...
       .cache()
 
-    val personalized = srcId.isDefined
     val src: VertexId = srcId.getOrElse(-1L)
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: Move up `src`

**Comment by dbtsai** [comment]
Seems that the implementation of personalized page rank doesn't follow the twitter's paper entirely when initializing the initial page rank. In the paper, only the source node should be activated. This PR addresses this issue. 

+cc @dwmclary @ankurdave @rxin @jegonzal for more feedback.

Thanks.

**Comment by dwmclary** [comment]
If I recall, we specifically decided against a conditional in the BSP function at that point because the branching might causes hotspots.  If that's still a concern, maybe @jegonzal can comment.  Otherwise, this looks good to me -- nice catch.

**Comment by dbtsai** [comment]
I think the branching should be fine since it's just initialization.

**Comment by dwmclary** [comment]
Yes, I agree, it shouldn't add overhead.

**Comment by jegonzal** [comment]
This is actually a pretty serious error since it could lead to mass being accumulated on unreachable sub-graphs.  The performance implications of the above branch should be negligible.

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala (hunk 2 above)
Comment (moustaki) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: Sorry @dbtsai not sure I understand this comment - where does src gets unboxed in that example?

**Comment by moustaki** [comment]
Jenkins, ok to test.

**Comment by dbtsai** [comment]
LGTM. Thanks, merged into master.

**Comment by moustaki** [comment]
Many thanks @dbtsai

**Comment by mengxr** [comment]
I backported this patch to branch-1.5 and 1.4.

**Pull request closed without merging**